src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,525 @@
+    // Put the variable in repls
+    repls.push_back((subvar_to_expr.begin())->first);
+
+    for (map<Node, Node>::iterator expr_it1 = subvar_to_expr.begin();
+         expr_it1 != subvar_to_expr.end();
+         ++expr_it1)
+    {
+      // Skip expr_it1->first, if it has been merged
+      if (merged.find(expr_it1->first) != merged.end())
+      {
+        continue;
+      }
+
+      Partition p;
+      vector<Node> subs;
+      vector<Node> part;
+      Node fst_var = subvar_to_var.find(expr_it1->first)->second;
+
+      part.push_back(fst_var);
+      subs.push_back(fst_var);
+      merged.insert(expr_it1->first);
+      p.d_var_to_subvar[fst_var] = expr_it1->first;
+      Node sub_expr1 =
+          (expr_it1->second)
+              .substitute(subs.begin(), subs.end(), repls.begin(), repls.end());
+
+      for (map<Node, Node>::iterator expr_it2 = subvar_to_expr.begin();
+           expr_it2 != subvar_to_expr.end();
+           ++expr_it2)
+      {
+        if (merged.find(expr_it2->first) != merged.end())
+        {
+          continue;
+        }
+        if ((expr_it1->second).getType() != (expr_it2->second).getType())
+        {
+          continue;
+        }
+        vector<Node> subs2;
+        Node snd_var = subvar_to_var.find(expr_it2->first)->second;
+
+        subs2.push_back(snd_var);
+        Node sub_expr2 =
+            (expr_it2->second)
+                .substitute(
+                    subs2.begin(), subs2.end(), repls.begin(), repls.end());
+        Trace("sym-dt") << "[sym-dt] Testing if " << sub_expr1
+                        << " is equivalent to " << sub_expr2 << endl;
+
+        if (sub_expr1 == sub_expr2)
+        {
+          Trace("sym-dt") << "[sym-dt] Merge variable " << fst_var
+                          << " with variable " << snd_var << endl;
+          merged.insert(expr_it2->first);
+          part.push_back(snd_var);
+          p.d_var_to_subvar[snd_var] = expr_it1->first;
+        }
+      }
+      p.d_subvar_to_vars[expr_it1->first] = part;
+      Trace("sym-dt") << "[sym-dt] Add a new partition after merging: " << endl;
+      printPartition(p);
+      partitions.push_back(p);
+    }
+  }
+}
+
+void SymmetryDetect::processSingletonPartitions(
+    vector<Partition>& partitions, unordered_set<Node, NodeHashFunction>& vars)
+{
+  Trace("sym-dt") << "[sym-dt] Start post-processing partitions with size = "
+                  << partitions.size() << endl;
+
+  // Mapping between the substitution variable to the actual variable
+  map<Node, Node> subvar_to_var;
+  // Mapping between the substitution variable to the actual expression
+  map<Node, Node> subvar_to_expr;
+  // Partitions that have 2 or more variables
+  vector<Partition> new_partitions;
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  // Collect singleton partitions: subvar_to_expr, subvar_to_var, and variables
+  while (part_it != partitions.end())
+  {
+    if (!(*part_it).d_var_to_subvar.empty()
```

Comment:
Isn't this check redundant? I.e. if you just checked `(*part_it).d_var_to_subvar.size() == 1` then non-emptiness would be implied, no?

Reply:
I was only checking (*part_it).d_var_to_subvar.size() before. However, somehow when (*part_it).d_var_to_subvar does not have any elements, (*part_it).d_var_to_subvar.size() would return some arbitrary number, which was a surprise to me. That is why I added to check if it is empty.

Reply:
Hmm, this should definitely not happen. Can you double-check?

Reply:
I have removed the checks. But I am not able to reproduce it now.